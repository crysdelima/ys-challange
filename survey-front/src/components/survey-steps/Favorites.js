import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Form, Checkbox } from 'semantic-ui-react';
import { debounce, isEmpty } from 'lodash';

import colorsList from '../../config/colorsList';

const onSaveWithDebounce = debounce((data, onSave, onDisableNextStep) => {
    onSave('favorites', data);
    
    const disable = Object.values(data).every(i => !isEmpty(i));
    return onDisableNextStep(!disable);
}, 300);

const Favorites = ({
    initialData,
    onDisableNextStep,
    onSave,
}) => {
    const [formData, setFormData] = useState({ ...initialData });

    useEffect(() => {
        const validFormData = Object.values(formData).every(i => !isEmpty(i))
        onDisableNextStep(!validFormData);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const onChange = e => {
        onSaveWithDebounce.cancel();
        const newFormData = {
            ...formData,
            [e.target.name]: e.target.value,
        };
        setFormData(newFormData);
        onSaveWithDebounce(newFormData, onSave, onDisableNextStep);
    };

    const onCheckboxChange = (e, { name, value, checked }) => {
        const colors = [...formData.colors];

        if (checked) {
            colors.push(value)
        } else {
            const valueIndex = colors.findIndex(item => item === value);
            colors.splice(valueIndex, 1);
        }
        return onChange({
            target: { name, value: colors }
        });
    }

    return (
        <Form>
            <Form.Input
                id='survey-identity-field-favorite-book'
                label='Favorite Book'
                name='book'
                onChange={onChange}
                required
                placeholder='Type your favorite book name'
                value={formData.book}
            />

            <Form.Field required>
                <label>Favorite Colors</label>
                {colorsList.map(item => (
                    <Form.Field key={item.value}>
                        <Checkbox
                            checked={formData.colors.includes(item.value)}
                            label={item.name}
                            name='colors'
                            onChange={onCheckboxChange}
                            value={item.value}
                        />
                    </Form.Field>
                ))}
            </Form.Field>
        </Form>
    );
};

Favorites.propTypes = {
    initialData: PropTypes.shape({
        book: PropTypes.string.isRequired,
        colors: PropTypes.array.isRequired,
    }).isRequired,
    onDisableNextStep: PropTypes.func.isRequired,
    onSave: PropTypes.func.isRequired,
};

export default Favorites;