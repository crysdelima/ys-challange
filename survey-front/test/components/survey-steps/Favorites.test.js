import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Favorites from '../../../src/components/survey-steps/Favorites';

describe('<Favorites />', () => {
    test('Should render correctly with inital value', () => {
        const onDisableNextStep = jest.fn();
        const wrapper = mount(
            <Favorites
                initialData={{ book: '', colors: [] }}
                onDisableNextStep={onDisableNextStep}
                onSave={() => {}}
            />
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('Should fill Favorite book input', done => {
        const onDisableNextStep = jest.fn();
        const onSave = (step, data) => {
            expect(step).toEqual("favorites")
            expect(data).toEqual({ book: "The book", colors: [] });
            expect(onDisableNextStep).toBeCalledWith(true);
            return done();
        };
        const wrapper = mount(
            <Favorites
                initialData={{ book: '', colors: [] }}
                onDisableNextStep={onDisableNextStep}
                onSave={onSave}
            />
        );
        wrapper.find('input[type="text"]').simulate('change', { target: {
            name: 'book', value: 'The book'
        }})
    });

    test('Should correctly handle color select', done => {
        const onDisableNextStep = jest.fn();
        const onSave = (step, data) => {
            expect(step).toEqual("favorites")
            expect(data).toEqual({ book: "", colors: ['green'] });
            expect(onDisableNextStep).toBeCalledWith(true);
            return done();
        };
        const wrapper = mount(
            <Favorites
                initialData={{ book: '', colors: [] }}
                onDisableNextStep={onDisableNextStep}
                onSave={onSave}
            />
        );
        wrapper.find('input[type="checkbox"]').at(1).simulate('change');
    });

    test('Should correctly handle color unselect', done => {
        const onDisableNextStep = jest.fn();
        const onSave = (step, data) => {
            expect(step).toEqual("favorites")
            expect(data).toEqual({ book: "", colors: ['blue'] });
            expect(onDisableNextStep).toBeCalledWith(true);
            return done();
        };
        const wrapper = mount(
            <Favorites
                initialData={{ book: '', colors: ['green', 'blue'] }}
                onDisableNextStep={onDisableNextStep}
                onSave={onSave}
            />
        );
        wrapper.find('input[type="checkbox"]').at(1).simulate('change');
    });

    test('Should correctly call onDisableNextStep with false when form is valid', done => {
        const onDisableNextStep = jest.fn();
        const onSave = (step, data) => {
            expect(step).toEqual("favorites")
            expect(data).toEqual({ book: "The book", colors: ['green'] });
            expect(onDisableNextStep).toBeCalledWith(false);
            return done();
        };
        const wrapper = mount(
            <Favorites
                initialData={{ book: '', colors: [] }}
                onDisableNextStep={onDisableNextStep}
                onSave={onSave}
            />
        );
        wrapper.find('input[type="text"]').simulate('change', { target: {
            name: 'book', value: 'The book'
        }})
        wrapper.find('input[type="checkbox"]').at(1).simulate('change');
    });
})
