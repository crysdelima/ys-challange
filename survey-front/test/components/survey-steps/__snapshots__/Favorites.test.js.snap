// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Favorites /> Should render correctly with inital value 1`] = `
<Favorites
  initialData={
    Object {
      "book": "",
      "colors": Array [],
    }
  }
  onDisableNextStep={
    [MockFunction] {
      "calls": Array [
        Array [
          true,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onSave={[Function]}
>
  <Form
    as="form"
  >
    <form
      className="ui form"
      onSubmit={[Function]}
    >
      <FormInput
        as={[Function]}
        control={[Function]}
        id="survey-identity-field-favorite-book"
        label="Favorite Book"
        name="book"
        onChange={[Function]}
        placeholder="Type your favorite book name"
        required={true}
        value=""
      >
        <FormField
          control={[Function]}
          id="survey-identity-field-favorite-book"
          label="Favorite Book"
          name="book"
          onChange={[Function]}
          placeholder="Type your favorite book name"
          required={true}
          value=""
        >
          <div
            className="required field"
          >
            <label
              htmlFor="survey-identity-field-favorite-book"
            >
              Favorite Book
            </label>
            <Input
              aria-describedby={null}
              id="survey-identity-field-favorite-book"
              name="book"
              onChange={[Function]}
              placeholder="Type your favorite book name"
              required={true}
              type="text"
              value=""
            >
              <div
                className="ui input"
              >
                <input
                  aria-describedby={null}
                  id="survey-identity-field-favorite-book"
                  name="book"
                  onChange={[Function]}
                  placeholder="Type your favorite book name"
                  required={true}
                  type="text"
                  value=""
                />
              </div>
            </Input>
          </div>
        </FormField>
      </FormInput>
      <FormField
        required={true}
      >
        <div
          className="required field"
        >
          <label>
            Favorite Colors
          </label>
          <FormField
            key="blue"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Blue"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="blue"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="blue"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="blue"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="blue"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Blue
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Blue
                          </label>,
                        }
                      }
                    >
                      <label>
                        Blue
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
          <FormField
            key="green"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Green"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="green"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="green"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="green"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="green"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Green
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Green
                          </label>,
                        }
                      }
                    >
                      <label>
                        Green
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
          <FormField
            key="yellow"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Yellow"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="yellow"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="yellow"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="yellow"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="yellow"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Yellow
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Yellow
                          </label>,
                        }
                      }
                    >
                      <label>
                        Yellow
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
          <FormField
            key="red"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Red"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="red"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="red"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="red"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="red"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Red
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Red
                          </label>,
                        }
                      }
                    >
                      <label>
                        Red
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
          <FormField
            key="orange"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Orange"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="orange"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="orange"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="orange"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="orange"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Orange
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Orange
                          </label>,
                        }
                      }
                    >
                      <label>
                        Orange
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
          <FormField
            key="purple"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Purple"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="purple"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="purple"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="purple"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="purple"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Purple
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Purple
                          </label>,
                        }
                      }
                    >
                      <label>
                        Purple
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
          <FormField
            key="gray"
          >
            <div
              className="field"
            >
              <Checkbox
                checked={false}
                label="Gray"
                name="colors"
                onChange={[Function]}
                type="checkbox"
                value="gray"
              >
                <div
                  className="ui checkbox"
                  onChange={[Function]}
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                >
                  <Ref
                    innerRef={
                      Object {
                        "current": <input
                          class="hidden"
                          name="colors"
                          readonly=""
                          tabindex="0"
                          type="checkbox"
                          value="gray"
                        />,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <input
                            class="hidden"
                            name="colors"
                            readonly=""
                            tabindex="0"
                            type="checkbox"
                            value="gray"
                          />,
                        }
                      }
                    >
                      <input
                        checked={false}
                        className="hidden"
                        name="colors"
                        readOnly={true}
                        tabIndex={0}
                        type="checkbox"
                        value="gray"
                      />
                    </RefFindNode>
                  </Ref>
                  <Ref
                    innerRef={
                      Object {
                        "current": <label>
                          Gray
                        </label>,
                      }
                    }
                  >
                    <RefFindNode
                      innerRef={
                        Object {
                          "current": <label>
                            Gray
                          </label>,
                        }
                      }
                    >
                      <label>
                        Gray
                      </label>
                    </RefFindNode>
                  </Ref>
                </div>
              </Checkbox>
            </div>
          </FormField>
        </div>
      </FormField>
    </form>
  </Form>
</Favorites>
`;
